### PR TITLE
[Storage] Support multiple files in Storage

### LINKED ISSUE #1070: [Storage] Upload multiple files to the same bucket with different directories
It would be great if sky storage can support uploading multiple files into the same bucket but different folder in the bucket. Currently, a workaround is to write another program to upload the files first, before running the sky yaml.

```python
for l_path in local_paths:
  subprocess.run(f'gsutil -m cp {l_path} gs://bucket-name/run-{run_id}', shell=True)
```

In yaml:
```yaml
file_mounts:
  remote_path: gs://bucket-name/run-{run_id}
```

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Closes #1070 and unblocks #1266.

Adds support for specifying multiple paths for upload to a Storage object using a list as a source:

```
  /dataset/:
    name: sky-multisource-storage\
    source: [~/mydir/myfile.txt, ~/datasets]
```

Tested:
- [x] `pytest test_smoke.py::TestStorageWithCredentials`
- [x] `sky launch examples/storage_demo.yaml`
- [x] Rendered docs locally 
- [x] All smoke tests

### CODE REVIEW

File: examples/storage_demo.yaml
Comment (Michaelvll) [inline examples/storage_demo.yaml:110]: For this line and L94, will we upload the `~/datasets` only once? As a future TODO, we may want to do a unique for the `(source, name)` pair.

File: sky/data/storage.py
Comment (Michaelvll) [inline sky/data/storage.py:348]: very small nit: `Union[Path, List[Path], None]`
https://google.github.io/styleguide/pyguide.html#3195-nonetype

File: sky/data/storage.py
Comment (Michaelvll) [inline sky/data/storage.py:1146]: Out of curiosity, why we are using `gsutil rsync` instead of `gsutil cp`? It seems that `gsutil cp` supports the following, which will significantly improve the uploading speed when multiple files are being uploaded. https://cloud.google.com/storage/docs/gsutil/commands/cp
```
cat filelist | gsutil -m cp -I gs://my-bucket
```

**Review by Michaelvll — COMMENTED** [review]
Thank you for adding this @romilbhardwaj ! It is very useful for many users and also for the auto-translation for the `file_mounts` in the spot jobs.
When I am reading the PR, and thinking how to implement the auto-translation with this. A problem come across, that how can user deal with files and folders with the same name. It becomes a bit tricky for the translation for spot, and I was wondering if the following is a valid (hidden) API for uploading a "list" of files:
```
source:  # Accepting a dict as well.
    dst_path/on_bucket/filename: source_path/file1.txt
    dst_path2: source_dir
```

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:348]: good catch!

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:519]: Ahh good point - fixed!

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:1146]: Great question - `gsutil cp` will force upload a file even if it already exists on the bucket every time the task is launched. `gsutil rsync` uploads only files which either don't exist on the bucket or have been modified locally. This reduces upload time and number of operations required to sync the storage object.

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:778]: done, thanks!

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:432]: I'm holding off on this for now because this would also affect the class constructor and thus the external interface. The eventual plan is to remove this arg from the Storage ctor, after which we should rename it.

File: examples/storage_demo.yaml
Comment (romilbhardwaj) [reply examples/storage_demo.yaml:110]: `gsutil rsync`/`aws s3 sync` will be invoked twice, but the contents will be uploaded only once. I think the motivation to use `~/datasets` at L110, L94 and L71 is to show that the same directory can be made available at the remote VM in different ways

File: sky/data/storage.py
Comment (Michaelvll) [reply sky/data/storage.py:1145]: Thanks for the explanation!
The reason I mentioned this is that for the salk users, we did observe that looping through all the files to be uploaded with `gsutil rsync` or `gsutil cp` was much slower than `cat filelist | gsutil -m cp -I gs://my-bucket/subdir`(10 min vs 3 min for more than 200 files). That is to say `sky spot launch` will be blocked by the uploading of the files.
It seems `gsutil cp` supports a `-n` option that avoids overwriting the files, but if that does not have the same semantic as `gsutil rsync`, I think it should be fine to keep the current way.

Comment (romilbhardwaj) [reply sky/data/storage.py:1145]: Oh that's a great point. It seems `gsutil cp -n` should do what we need. Let me try it out and update the PR if it works

Comment (romilbhardwaj) [reply sky/data/storage.py]: Thanks for the pointer @Michaelvll, `gsutil cp -n` indeed works similar to rsync and calling `cat files | gsutil -m cp -n -I` once is faster than calling `gsutil rsync` multiple times. Switched our implementation. PTAL

**Comment by romilbhardwaj** [comment]
Thanks for the pointer @Michaelvll, `gsutil cp -n` indeed works similar to `gsutil rsync` and calling `cat files | gsutil -m cp -n -I` once is faster than calling `gsutil rsync` multiple times when uploading many files. Switched our implementation. PTAL

**Review by Michaelvll — APPROVED** [review]
Thanks for the quick fixes @romilbhardwaj! LGTM!

A side note: please correct me if I am wrong, but seems both `gsutil rsync` and `gsutil cp -n` will not consider the content change in a file, i.e. adding lines in the file locally will not update the existing file on the bucket. This behavior could be a bit surprising, and worth a comment. Wdyt?

File: docs/source/reference/storage.rst
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:165]: I think we can make this more concise.
(reword)
```
The basenames of each path in the source list are copied recursively to the root of the bucket.
```

File: docs/source/reference/storage.rst
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:263]: Will the dictionary of sources be in another PR?

File: docs/source/reference/storage.rst
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:264]: Grammar: `or a list of local paths. Data is uploaded to the cloud to ....`

File: examples/storage_demo.yaml
Comment (michaelzhiluo) [inline examples/storage_demo.yaml:40]: Inconsistency, lets keep s3 and gcs lower case?

File: sky/utils/schemas.py
Comment (michaelzhiluo) [inline sky/utils/schemas.py:86]: how do you specify a Union[string, array of strings]?

**Comment by romilbhardwaj** [comment]
> A side note: please correct me if I am wrong, but seems both `gsutil rsync` and `gsutil cp -n` will not consider the content change in a file, i.e. adding lines in the file locally will not update the existing file on the bucket. This behavior could be a bit surprising, and worth a comment. Wdyt?

I just tested this - `gsutil rsync` considers content changes when uploading file, whereas `gsutil cp -n` just looks at same filename. 

This presents a functionality vs performance tradeoff for us. 
1. `gsutil cp -n` is faster (3 min for more than 200 files), but does not update the file in the bucket if its contents change (e.g. has lines).
2. `gsutil rsync` is slower (10 min for more than 200 files), but updates the file in the bucket if it has local changes.

I'm leaning towards 2 because if the user wants to update a file in the bucket, the only option with 1. is to delete and bucket and upload again or upload the file with a new file name (and the older one does not get deleted). 

Another option is to do 1 when creating the storage object for the first time, and using 2 for subsequent calls.

**Comment by Michaelvll** [comment]
> I'm leaning towards 2 because if the user wants to update a file in the bucket, the only option with 1. is to delete and bucket and upload again or upload the file with a new file name (and the older one does not get deleted).
>
>Another option is to do 1 when creating the storage object for the first time, and using 2 for subsequent calls.

Ahh, good catch! Then sticking with option 2 sounds good to me for the existing bucket. "only option 1 for creating the storage object for the first time" seems not applicable to the salk user's use case, since they always upload the files to the same bucket, so even we have that, it will always choose the `rsync`.

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:165]: great point! while this is concise, I feel it becomes harder to read for a non-systems/os audience. I've kept the original and added this before it.

File: docs/source/reference/storage.rst
Comment (romilbhardwaj) [reply docs/source/reference/storage.rst:264]: Fixed!

File: examples/storage_demo.yaml
Comment (romilbhardwaj) [reply examples/storage_demo.yaml:40]: hmm in this doc, we consistently use `S3/GCS` when referring to the services, and lowercase `s3/gcs` when used in urls/paths

File: sky/utils/schemas.py
Comment (romilbhardwaj) [reply sky/utils/schemas.py:86]: I believe it's [type1, type2] [source](https://cswr.github.io/JsonSchema/spec/multiple_types/)

**Comment by romilbhardwaj** [comment]
Based on our discussions above, I've reverted to using `rsync` instead of `cp`. I have tried to make rsync faster by 1. Grouping rsync calls and 2. Running them in parallel. This should be faster than sequentially running rsync for all files. PTAL!

File: sky/data/storage.py
Comment (romilbhardwaj) [inline sky/data/storage.py:1202]: This is unused in this PR, but will be required in the next PR which allows source to be a dict since rsync doesn't allow renaming at destination.

File: sky/utils/schemas.py
Comment (michaelzhiluo) [reply sky/utils/schemas.py:86]: In that case, is it possible to replace `array` with:

```
{
  "type": "array",
  "items": {
    "type": "string"
  }
}
```

File: examples/storage_demo.yaml
Comment (michaelzhiluo) [reply examples/storage_demo.yaml:40]: Sounds good!

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:518]: If it a single file, can we automatically convert it to a list instead?

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:814]: `if self.source is not None and isinstance(self.source.str)` (want to avoid singular nested if statements)

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:861]: Can we abstract this to `AbstractStore`  : `batch_sync(...)`

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:936]: We should use existing code in backend.py that can run several commands in parallel: https://github.com/skypilot-org/skypilot/blob/2336d75fc21d04329634b89d0424d6dceaad698b/sky/backends/cloud_vm_ray_backend.py#L2041

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:1256]: The structures for both methods `batch_aws_sync` and `batch_gsutil_rsync` are very similar. Can we group it together in AbstractStore?

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:425]: fixed!

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:814]: fixed!

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:936]: In the refactor, it's not possible to reuse because this code chunk now requires using pool.starmap and a custom number of workers

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:861]: Moved it to data_utils.py!

**Comment by romilbhardwaj** [comment]
Ready for another look @michaelzhiluo!

File: sky/data/data_utils.py
Comment (Michaelvll) [inline sky/data/data_utils.py:163]: Will this correctly fail the other threads when one child fails? Just want to make sure that there will be no leaked thread.
Here is the reason we switched to `p.imap` in `subprocess_utils`.
https://stackoverflow.com/questions/25790279/python-multiprocessing-early-termination

File: sky/data/storage.py
Comment (Michaelvll) [reply sky/data/storage.py]: Will this correctly fail the other threads when one child fails? Just want to make sure that there will be no leaked thread.
Here is the reason we switched to `p.imap` in `subprocess_utils`.
https://stackoverflow.com/questions/25790279/python-multiprocessing-early-termination

Comment (romilbhardwaj) [reply sky/data/storage.py]: @Michaelvll - yes, seems like if an exception is raised, `pool.starmap` correctly terminates the ThreadPool and threads don't leak. The error in the stackoverflow link happens only if `sys.exit` is called in the run function. I tested with this snippet:

```
import subprocess
import sys
from multiprocessing import pool


def run_fn(command, message):
    print(message)
    if '1' in command:
        command += '; asdf'  # Inject error
    with subprocess.Popen(command,
                          stderr=subprocess.PIPE,
                          stdout=subprocess.DEVNULL,
                          shell=True) as process:
        stderr = []
        while True:
            line = process.stderr.readline()
            if not line:
                break
            str_line = line.decode('utf-8')
            stderr.append(str_line)
        returncode = process.wait()
        if returncode != 0:
            # sys.exit(0) # p.map or starmap is problematic only if you raise sys.exit here
            raise RuntimeError(f'Command {command} failed.')


commands = [[f"echo {i}", f"Command{i}"] for i in range(0, 10)]
with pool.ThreadPool(processes=16) as p:
    res = p.starmap(run_fn, commands)
```

File: sky/data/storage.py
Comment (michaelzhiluo) [reply sky/data/storage.py:516]: As discussed in meeting, not a good idea.

File: sky/data/data_utils.py
Comment (michaelzhiluo) [inline sky/data/data_utils.py:144]: Nit: It seems that `group_files_by_dir` is only called by this method. Make it `_group_files_by_dir`

File: tests/test_smoke.py
Comment (michaelzhiluo) [inline tests/test_smoke.py:1113]: I just tested this out, it works, awesome!

File: tests/test_smoke.py
Comment (michaelzhiluo) [inline tests/test_smoke.py:1066]: It is a bit awkward have an unspecified requirement for the suffix after the bucket name. I.e. is the suffix `/root_path` or `root_path`?

File: tests/test_smoke.py
Comment (michaelzhiluo) [inline tests/test_smoke.py:1120]: Look above, this was why I was a bit confused.

File: sky/data/data_utils.py
Comment (michaelzhiluo) [inline sky/data/data_utils.py:114]: For typing, to specify a function, need to also type out the types of arguments the functions takes (i.e. `Callable[int]`)

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:930]: Has a scalability test been tried for this? (i.e. what if we had 32 imagenet folders to upload to the bucket, does it kill the computer?) I checked online and it says that `aws s3 sync` by default uses 10 threads. 10*32 = 320 threads.

Comment (michaelzhiluo) [reply sky/data/storage.py:930]: It might be good to try a larger version in a smoke test, but I'm fine with leaving it out for now.

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:900]: `get_file_sync_command` and `get_dir_sync_command` can be united under a single function. (in that function it will check if the uploaded file/folder is a file/folder). This would make `parallel_upload` a lot simpler.

File: sky/data/storage.py
Comment (michaelzhiluo) [inline sky/data/storage.py:1215]: Ditto.

**Review by michaelzhiluo — APPROVED** [review]
Nice abstractions that are moved to `data_utils.py` @romilbhardwaj ! The method in `batch_aws/gcs_sync` is much cleaner now. Good to merge after these comments are addressed.

File: tests/test_smoke.py
Comment (romilbhardwaj) [reply tests/test_smoke.py:1066]: Thanks! I've made it more explicit now

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:900]: Ah, the reason I did it this way was because that would result in code duplication in S3Store and GCStore classes. E.g., both would implement something like:

```python
def get_sync_command(source_path_list):
    commands = []
    grouped_files, dirs = data_utils.group_files_by_dir(source_path_list)
    # Generate file upload commands
    for dir_path, file_names in grouped_files.items():
        sync_command = get_file_sync_command(dir_path, file_names)
        commands.append(sync_command)
    # Generate dir upload commands
    for dir_path in dirs:
        if create_dirs:
            dest_dir_name = os.path.basename(dir_path)
        else:
            dest_dir_name = ''
        sync_command = get_dir_sync_command(dir_path, dest_dir_name)
        commands.append(sync_command)
    return commands
```

I figured it's cleaner to extract this method into data_utils.py instead of duplicating it in the two classes or implementing in a parent class. lmk if you feel strongly otherwise, I can change this

File: sky/data/storage.py
Comment (romilbhardwaj) [reply sky/data/storage.py:930]: Great point - I just tried with 20 file_mounts, seemed to work okay on my 2018 x1 yoga (i7-8650u, 16gb RAM).

**Comment by romilbhardwaj** [comment]
Thanks for the review @michaelzhiluo! I've resolved your comments and left a note on one comment. Re-running tests now. If everything looks good I'll go ahead and merge.

File: sky/data/storage.py
Comment (michaelzhiluo) [reply sky/data/storage.py:930]: @romilbhardwaj After this PR, good idea to file a Git issue for scalability tests for Sky Storage!

File: sky/data/storage.py
Comment (michaelzhiluo) [reply sky/data/storage.py:900]: Hmm yeah the former options seems better than. Its less LOC. Sounds good!

File: sky/utils/schemas.py
Comment (michaelzhiluo) [reply sky/utils/schemas.py:91]: @romilbhardwaj Double check this one, does this work?

**Comment by michaelzhiluo** [comment]
@romilbhardwaj PR looks good to go! Just address the two comments here and feel free to merge^^

https://github.com/skypilot-org/skypilot/pull/1311#discussion_r1029995926

https://github.com/skypilot-org/skypilot/pull/1311#discussion_r1029999097

**Comment by romilbhardwaj** [comment]
Awesome! I fixed the schema to specify types for the array in too. Merging now.

**Pull request merged**
